### DISCUSSION TITLE
Redirect URL for oauth flow

### DISCUSSION THREAD

**Discussion by dev-mash** [tauri-apps/tauri#8554] [General]
I'm using oauth2 to authenticate users. This works fine with windows. My redirect URL is  like https://tauri.localhost. Now I started to look into linux and found that the schema is tauri://localhost. But this redirect URL fails with following error:

"Redirection to URL with a scheme that is not HTTP(S)"

Is it possible to also have https://tauri.localhost on linux or is there any other solution to this problem?

**Comment by FabianLars** [comment]
You could make your app run on `http://localhost:{port}` via https://github.com/tauri-apps/plugins-workspace/tree/v1/plugins/localhost but this is only recommended if there's really no other way since it exposes your frontend to the whole system (if that matters for you).

Generally i'd recommend spawning a seperate localhost server on the rust side that only handles the redirect request, for non rust users i created a small plugin here (still missing docs) <https://github.com/FabianLars/tauri-plugin-oauth>. I created the plugin initially to be able to run the oauth flow in the browser for oauth services that don't support deep links (because i really hate in-app logins as a user 😅), however it does work for in-app flows too, you'll just have to manually trigger an url change if you register a call to the localhost server.

Edit: I forgot to give a straight answer to your last question: no it's not possible to use `http(s)://tauri.localhost` on linux or macos.

**Reply by TylerGrayStudent** [reply]
Would you have an example of how to use this? I'm really new to rust and would appreciate it a ton!

Edit: I see the examples. Silly me.

**Comment by dev-mash** [comment]
Thanks for your answer @FabianLars. Other that stated before, I think my provider actually supports deeplinks (i.e. yourapp:// as a redirect URL). So I think I would need to intercept the webviews redirect to yourapp:// and do the token fetching via a implementation in rust. Once that was successful, I redirect back to the app, i.e.  tauri://localhost

I'm actually new to rust, so I tried to avoid to do the token fetching via rust. On windows I implemented the token fetching (a http POST request) in javascript. I think I could try to call my current javascript implementation from the rust interceptor.

I tried to adapt the code from this example:
https://github.com/FabianLars/tauri-plugin-deep-link/blob/main/example/main.rs

I started the app in dev mode (npm run tauri dev) and simulated a redirect via following javascript in the dev console:
window.location.replace("testapp://localhost");

I was expecting to see "Hello World" printed. But it did not. Do you have any idea why this is not called?

Thanks in advance &
Best regards,
Manuel

My code in main.rs is similar to this:


```
println!("Starting RUST GREETINGS World!");    
tauri_plugin_deep_link::prepare("de.fabianlars.deep-link-test");
    tauri::Builder::default()
    .setup(|app| {
          tauri_plugin_deep_link::register(
            "testapp",
            move |request| {
              println!("Hello World!");
              dbg!(&request);
//               handle.emit_all("scheme-request-received", request).unwrap();
            },
          )
          .unwrap(/* If listening to the scheme is optional for your app, you don't want to unwrap here. */);

          Ok(())
        })
        .invoke_handler(tauri::generate_handler![greet])
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
```

I get the following error in the dev console:
[Error] Unhandled Promise Rejection: Scope not defined for window `main` and URL `testapp://localhost`. See https://tauri.app/v1/api/config/#securityconfig.dangerousremotedomainipcaccess and https://docs.rs/tauri/1/tauri/scope/struct.Ipc...

**Reply by FabianLars** [reply]
On which OS are you testing this? Also just windows?
What happens if you navigate to that url in a actual browser window?
What happens if you call the deep link while the app is not running? Does it open still it?
> I was expecting to see "Hello World" printed. But it did not. Do you have any idea why this is not called?

Just to be sure, you looked into the terminal where `tauri dev` runs, right?

> I get the following error in the dev console:
[Error] Unhandled Promise Rejection: Scope not defined for window main and URL testapp://localhost. See https://tauri.app/v1/api/config/#securityconfig.dangerousremotedomainipcaccess and https://docs.rs/tauri/1/tauri/scope/struct.Ipc...

That sounds like the system didn't register the deep link and actually tries to navigate to that url 🤔

**Reply by dev-mash** [reply]
I have tested this on Ubuntu 22.04.3 LTS
1. starting the app (npm run tauri dev)
2. open browser dev console
3. open my redirect url --> window.location.href = "testapp://localhost"
4. app shows: "The URL can’t be shown"
5. Error in dev console: [Error] Unhandled Promise Rejection: Scope not defined for window `main` and URL `testapp://localhost`. See https://tauri.app/v1/api/config/#securityconfig.dangerousremotedomainipcaccess and https://docs.rs/tauri/1/tauri/scope/struct.Ipc...

Independent of if the app is running, when I open testapp://localhost in my browser (firefox) this navigates to  google-search: https://www.google.com/search?q=testapp%3A%2F%2Flocalhost

Yes I looked in the terminal where I already see the initial test-printout from the main method ("Starting RUST GREETINGS World!").

I use tauri-plugin-deep-link = "0.1.2", maybe this is not working anymore?

**Reply by FabianLars** [reply]
The plugin should still work, but Linux is often a pain with things like these.

Let's try this: In the src-tauri folder run `cargo add pretty_env_logger` and add this to the top of the main function in main.rs
```rs
pretty_env_logger::init();
```
Example: https://github.com/FabianLars/mw-toolbox/blob/main/gui/src-tauri/src/main.rs#L34

Then, try again and check for new logs in the terminal.

**Reply by dev-mash** [reply]
Thanks @FabianLars . I see the following logs:
Starting RUST GREETINGS World!
 ERROR tauri_plugin_deep_link::platform_impl > Error creating socket listener: No such file or directory (os error 2)

**Reply by FabianLars** [reply]
hmm, that one doesn't make sense unless your system doesn't have a `/tmp` dir but ubuntu should have it. Can you check if there's a <identifier>-deep-link.sock file in the /tmp dir? (before you run tauri dev _and_ while it's running)

**Reply by dev-mash** [reply]
Yes I can see this file in the /tmp folder when running the app:
de.fabianlars.deep-link-test-deep-link.sock

Content of the file shows: cat de.fabianlars.deep-link-test-deep-link.sock 
cat: de.fabianlars.deep-link-test-deep-link.sock: No such device or address

**Reply by FabianLars** [reply]
just in case: make sure to change that identifier. it must be unique so using the one from the example is not a good idea.

Anyway, the behavior seems to be correct then so i'm a bit confused what's going on. Also, ignore the `cat` error, i think that's normal because it's not an actual file.

Can you check the output of `xdg-mime query default x-scheme-handler/<your-protocol>` then?

**Reply by dev-mash** [reply]
This is what I get:
manuel@Manuel:/tmp$ xdg-mime query default x-scheme-handler/testapp
tasks-handler.desktop

**Reply by FabianLars** [reply]
that's also correct and should mean that it's registered in your system 🤔 

Can you post the output of the `tauri info` command? I'll try to repro it on my end tomorrow.

**Comment by dev-mash** [comment]
Thanks for your support:

npm run tauri info

> myapp@0.0.0 tauri
> tauri info


[✔] Environment
    - OS: Ubuntu 22.04 X64
    ✔ webkit2gtk-4.0: 2.42.4
    ✔ rsvg2: 2.52.5
    ✔ rustc: 1.75.0 (82e1608df 2023-12-21)
    ✔ cargo: 1.75.0 (1d8b05cdd 2023-11-20)
    ✔ rustup: 1.26.0 (5af9b9484 2023-04-05)
    ✔ Rust toolchain: stable-x86_64-unknown-linux-gnu (default)
    - node: 18.12.1
    - npm: 8.19.2

[-] Packages
    - tauri [RUST]: 1.4.1
    - tauri-build [RUST]: 1.4.0
    - wry [RUST]: 0.24.3
    - tao [RUST]: 0.16.2
    - @tauri-apps/api [NPM]: 1.5.3
    - @tauri-apps/cli [NPM]: 1.5.9

[-] App
    - build-type: bundle
    - CSP: unset
    - distDir: ../dist/myapp-ui
    - devPath: http://localhost:1420/tasks
    - framework: Angular
    - bundler: Webpack

**Reply by dev-mash** [reply]
Hi @FabianLars, did you had a chance to check this from your end?
Best regards,
Manuel

**Reply by dev-mash** [reply]
I gave it another try and I noticed a error in logs:
ERROR tauri_plugin_deep_link::platform_impl > Error creating socket listener: No such file or directory (os error 2)

Apparently this happens during:
tauri_plugin_deep_link::prepare("com.test.deep-link-test");

Can you help me to further troubleshoot this, @FabianLars ?

Best regards,
Manuel

**Reply by FabianLars** [reply]
That error doesn't tell us much, or really, it's not necessarily an error actually. I just messed up the wording a bit here. Is your project public? Or can you create a minimal reproduction repo we can try?

**Comment by dev-mash** [comment]
> That error doesn't tell us much, or really, it's not necessarily an error actually. I just messed up the wording a bit here. Is your project public? Or can you create a minimal reproduction repo we can try?

I setup a new project. I took the example from your project: https://github.com/FabianLars/tauri-plugin-deep-link/blob/main/example/main.rs

This works on my windows machine, but its not working on my linux machine. Browser console shows this (when trying to redirect to my-scheme://test.de):
[Error] Failed to load resource: The URL can’t be shown (test.de, line 0)
[Error] Unhandled Promise Rejection: Scope not defined for window `main` and URL `my-scheme://test.de`. See https://tauri.app/v1/api/config/#securityconfig.dangerousremotedomainipcaccess and https://docs.rs/tauri/1/tauri/scope/struct.Ipc...

I added the dependency to the cargo.toml:
tauri-plugin-deep-link = "=0.1.2"

and added this in the main.rs

  tauri_plugin_deep_link::prepare("com.test.deep-link-test");

and in the setup method of tauri builder:
      tauri_plugin_deep_link::register(
        "my-scheme",
        move |request| {
          println!("Hello Handler");
          dbg!(&request);
//           handle.emit_all("scheme-request-received", request).unwrap();
        },
      )
      .unwrap(/* If listening to the scheme is optional for your app, you don't want to unwrap here. */);

**Reply by FabianLars** [reply]
can you check what happens if you write this in your terminal?
```
xdg-mime query default x-scheme-handler/my-scheme
```

**Reply by dev-mash** [reply]
It gives following output:
tasks-handler.desktop

My system details I also provided last time (npm run tauri info)

I now updated from tauri 1.4 to 1.5 and set dangerousUseHttpScheme in the tauri config to true. But this also did not solve the problem.
